tskv/src/file_system/file/cursor.rs

```diff
@@ -73,18 +83,21 @@ impl FileCursor {
         .ok_or_else(|| Error::new(ErrorKind::InvalidInput, "underflow or overflow during seek"))?;
         Ok(self.pos)
     }
-}
 
-impl From<AsyncFile> for FileCursor {
-    fn from(file: AsyncFile) -> Self {
-        FileCursor { file, pos: 0 }
+    pub fn len(&self) -> u64 {
+        self.file.len()
+    }
+    pub fn is_empty(&self) -> bool {
+        self.file.is_empty()
     }
 }
 
-impl Deref for FileCursor {
-    type Target = AsyncFile;
-
-    fn deref(&self) -> &Self::Target {
-        &self.file
+impl From<AsyncFile> for FileCursor {
+    fn from(file: AsyncFile) -> Self {
+        FileCursor {
+            file,
+            pos: 0,
+            buf: Vec::with_capacity(BUFFER_SIZE),
+        }
     }
```

Comment:
Please try this code:

```rust
#[cfg(test)]
mod test {
    use std::io::SeekFrom;
    use std::path::Path;

    use super::FileCursor;
    use crate::file_system::file::IFile;
    use crate::file_system::file_manager;

    #[tokio::test]
    async fn test() {
        let test_dir = "/tmp/test/cursor/";
        let _ = std::fs::remove_dir_all(test_dir).unwrap();
        std::fs::create_dir_all(test_dir).unwrap();
        let file_path = Path::new(test_dir).join("test.txt");

        {
            let file = file_manager::create_file(&file_path).await.unwrap();
            let mut cur = FileCursor::from(file);
            assert_eq!(cur.pos, 0);
            assert_eq!(cur.len(), 0);

            // Seek to pos: 8
            cur.seek(SeekFrom::Start(8_u64)).await.unwrap();
            assert_eq!(cur.pos, 8);
            assert_eq!(cur.pos(), 8);
            assert_eq!(cur.len(), 0);

            // Write at pos: 8, data: [1,1,1,1], pos_after: 12
            cur.write(&[1_u8; 4]).await.unwrap();
            assert_eq!(cur.pos, 8);
            assert_eq!(cur.pos(), 12);
            assert_eq!(cur.len(), 0);
            assert_eq!(cur.buf, vec![1_u8; 4]);

            // Write at pos: 12, data: [2,2,2,2], pos_after: 16
            cur.write(&[2_u8; 4]).await.unwrap();
            assert_eq!(cur.pos, 8);
            assert_eq!(cur.pos(), 16);
            assert_eq!(cur.len(), 0);
            assert_eq!(cur.buf, vec![1, 1, 1, 1, 2, 2, 2, 2]);

            // Seek to pos: current(4), pos_after: 20
            cur.seek(SeekFrom::Current(4)).await.unwrap();
            assert_eq!(cur.pos, 20);
            assert_eq!(cur.pos(), 20);
            assert_eq!(cur.len(), 0);
            assert_eq!(cur.buf, Vec::<u8>::new());

            // Seek to pos: end(4), pos_after: 24
            cur.seek(SeekFrom::End(4)).await.unwrap();
            assert_eq!(cur.pos, 24);
            // assert_eq!(cur.pos, 4);
            assert_eq!(cur.pos(), 24);
            // assert_eq!(cur.pos(), 4);
            assert_eq!(cur.len(), 0);
            assert_eq!(cur.buf, Vec::<u8>::new());

            // Write at pos: 24, data: [3,3,3,3], pos_after: 28
            cur.write(&[3_u8; 4]).await.unwrap();
            cur.try_flush(0).await.unwrap();
            drop(cur);
        }
        {
            let file = file_manager::open_file(&file_path).await.unwrap();
            let mut buffer = vec![0_u8; 64];
            let data_len = file.read_at(0, &mut buffer).await.unwrap();
            #[rustfmt::skip]
            assert_eq!(
                &buffer[..data_len],
                &[
                    0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 2,
                    0, 0, 0, 0, 0, 0, 0, 0, 3, 3, 3, 3,
                ]
            );
        }
    }
}
```

```sh
thread 'file_system::file::cursor::test::test' panicked at 'assertion failed: `(left == right)`
  left: `[0, 0, 0, 0, 3, 3, 3, 3, 1, 1, 1, 1, 2, 2, 2, 2]`,
 right: `[0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 2, 0, 0, 0, 0, 0, 0, 0, 0, 3, 3, 3, 3]`', tskv/src/file_system/file/cursor.rs:173:13
```

Reply:
This use case would not pass in the original code either. Since our file.len is not dynamically updated, in order to prevent misunderstanding for users, I removed support for Seek::End in the seek operation.